athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/GcsMetadataHandler.java

```diff
@@ -0,0 +1,291 @@
+     * @return A ListSchemasResponse which primarily contains a Set<String> of schema names and a catalog name
+     * corresponding the Athena catalog that was queried.
+     */
+    @Override
+    public ListSchemasResponse doListSchemaNames(BlockAllocator allocator, ListSchemasRequest request) throws Exception
+    {
+        return super.doListSchemaNames(allocator, request, DB_FILTER);
+    }
+
+    /**
+     * Used to get the list of tables that this source contains.
+     *
+     * @param allocator Tool for creating and managing Apache Arrow Blocks.
+     * @param request   Provides details on who made the request and which Athena catalog and database they are querying.
+     * @return A ListTablesResponse which primarily contains a List<TableName> enumerating the tables in this
+     * catalog, database tuple. It also contains the catalog name corresponding the Athena catalog that was queried.
+     */
+    @Override
+    public ListTablesResponse doListTables(BlockAllocator allocator, final ListTablesRequest request) throws Exception
+    {
+        return super.doListTables(allocator,
+                new ListTablesRequest(request.getIdentity(), request.getQueryId(),
+                        request.getCatalogName(), request.getSchemaName(), null, UNLIMITED_PAGE_SIZE_VALUE),
+                TABLE_FILTER);
+    }
+
+    /**
+     * Used to get definition (field names, types, descriptions, etc...) of a Table.
+     *
+     * @param blockAllocator Tool for creating and managing Apache Arrow Blocks.
+     * @param request        Provides details on who made the request and which Athena catalog, database, and table they are querying.
+     * @return A GetTableResponse which primarily contains:
+     * 1. An Apache Arrow Schema object describing the table's columns, types, and descriptions.
+     * 2. A Set<String> of partition column names (or empty if the table isn't partitioned).
+     * 3. A TableName object confirming the schema and table name the response is for.
+     * 4. A catalog name corresponding the Athena catalog that was queried.
+     */
+    @Override
+    public GetTableResponse doGetTable(BlockAllocator blockAllocator, GetTableRequest request) throws Exception
+    {
+        GetTableResponse response = super.doGetTable(blockAllocator, request);
+        //check whether schema added by user
+        //return if schema present else fetch from files(dataset api)
+        if (response != null && response.getSchema() != null && (response.getSchema().getFields().stream().count() - response.getPartitionColumns().stream().count()) > 0) {
+            return response;
+        }
+        else {
+            LOGGER.warn("Fetching schema from google cloud storage files");
+            //fetch schema from GCS in case user doesn't define it in glue table
+            //this will get table(location uri and partition details) without schema metadata
+            Table table = GcsUtil.getGlueTable(request.getTableName(), glueClient);
```

Comment:
I am confused with else clause statement here. 

As the comment mentioned (below), we are fetching schema from GCS in case user doesn't define in glue.
`   //fetch schema from GCS in case user doesn't define it in glue table

//this will get table(location uri and partition details) without schema metadata`

1.  I thought Glue is mandatory?
2. we are still calling table from glue  `GcsUtil.getGlueTable` and doing post processing here in else clause. this logic is pretty much identical with `super.doGetTable(blockAllocator, request);` logic. 

Questions: 
1.  What is the intention for this else clause.
2. Do we really need this else clause if they end up doing the same thing (call glue and do generate getable GetTableResponse)

Reply:
Simply this logic for GcsMetadataHandler.doGetTable with GlueMetadataHandler.doGetTable.

Reply:
Figure it out the logic, we get necessary information from glue to invoke datasource API to retrieve schema information.  Resolving this converseation